### DISCUSSION TITLE
Underline, italic, bold, font type and size properties using Textbox

### DISCUSSION THREAD

**Discussion by Lautarob1** [TomSchimansky/CustomTkinter#2074] [Q&A]
In the site/webpage https://customtkinter.tomschimansky.com/documentation/widgets/textbox I couldn't see how to set the properties for lines within CTkTextboxes. In plain tkinter, (tk.Text) you do (for example):
        txt_info.tag_configure("bold", font=("Arial", 12, "bold"))
        txt_info.tag_configure("italic", font=("Arial", 12, "italic"))
        txt_info.tag_configure("underline", underline=True)
        txt_info.tag_configure("center", justify="center")
        txt_info.tag_configure("multi_sty1", font=("Arial", 12, "bold italic"), underline=True, justify="center")
Then, you apply the styles (underline, multi_sty1 etc) in each line. But I haven't seen information on how to do the same using CTK textbox widget

**Comment by zSynctic** [comment]
Use txt_info.configure(font=("Arial", 12))

**Comment by Lautarob1** [comment]
Thanks! Now, it seems to me that this configure the whole widget (every text in it). What I was looking for is how to configure (differently) each line of the text, as tk.Text allows to doLautaro Barrera BOn Nov 16, 2023, at 4:01 AM, Synctic ***@***.***> wrote:﻿
Use txt_info.configure(font=("Arial", 12))

**Reply by zSynctic** [reply]
So you mean changing the font of specific words? or how to make some bold?

**Reply by zSynctic** [reply]
If you meant how to make a text bold or italic or underlined then use this: `txt_info.configure(font=("Arial", 12, 'bold'))` you could change bold to italic or underline

**Comment by Lautarob1** [comment]
Correct!For example underline and center one specific line, make another using a different font etc. even change the style for a couple of words inside a line, just as tk.Text allowsLautaro Barrera BOn Nov 16, 2023, at 8:53 AM, Synctic ***@***.***> wrote:﻿
So you mean changing the font of specific words? or how to make some bold?

**Reply by zSynctic** [reply]
I just tested and yes there is a tag_add also tag_config in a CTkTextbox so try that also try this way in the link and let me know if it works or not: https://stackoverflow.com/questions/52679038/can-i-set-different-fonts-for-different-lines-in-a-tkinter-text-widget

**Comment by Lautarob1** [comment]
By the way, and probably this is not the proper way to ask this, but I’m having troubles using ctk in MacOS (segmentation errors). I don’t think this is directly related to ctk but to Tkinter in general. Do you know a good source of information to troubleshoot this issue? Lautaro Barrera BOn Nov 16, 2023, at 8:53 AM, Synctic ***@***.***> wrote:﻿
So you mean changing the font of specific words? or how to make some bold?

**Reply by zSynctic** [reply]
Sorry but I have no idea about this matter since I don't use MacOS but maybe I can find a way to solve your problem so can you explain it more

**Comment by zSynctic** [comment]
Did it work?

**Comment by Lautarob1** [comment]
Thanks!

 

Perhaps I missed something here: “I just tested and yes there is a tag_add also tag_config in a CTkTextbox”

 

This is the test I did:

 

        self.txt_Usageinfo = ctk.CTkTextbox(self.frame_content0, wrap=tk.WORD, width=540, height=15)

        self.txt_Usageinfo.grid(row = 6, column = 0, columnspan=3, padx = (50,10), pady=(5,10)) #, sticky = 'w'

 

        lin1 = "Important Usage Information"

        self.txt_Usageinfo.tag_add("lin1","1.0", "1.end")

        self.txt_Usageinfo.tag_configure("lin1", font=("Arial", 16, "bold"))

 

        lin2 = "Second line of this test about usage info"

        self.txt_Usageinfo.tag_add("lin2","3.0", "3.end")

        self.txt_Usageinfo.tag_configure("lin2", font=("Arial", 12, "Italic"))

        

        self.txt_Usageinfo.insert("end", lin1+"\n")

        self.txt_Usageinfo.insert("end", lin2+"\n")

 

AttributeError: 'CTkTextbox' object has no attribute 'tag_configure'. Did you mean: '_configure'?

 

I know that tk.Text allows to use different fonts and styles in the text placed, but the ctk manual was silent and while researching on the internet, I came across with a post that says that only can be done at the global level (meaning one style for the whole text). I’m not saying that this is necessarily correct, but as the manual is silent on how to do it, I assumed so. 

 

The reason for not using tk.Text is (obviously) it dated appearance… As Opposed CTK widgets looks modern and neat. A big difference!! 

 

Thanks!

**Reply by zSynctic** [reply]
Np, thankfully it worked. if you liked my answer, can you mark it as an answer if you don't mind. thanks if you have any other questions let me know.

**Reply by Lautarob1** [reply]
Sorry, probably I have a dated version of Customtkinter (which I usually import as ctk).  In my version, I get the error shown. (AttributeError: 'CTkTextbox' object has no attribute 'tag_configure'. Did you mean: '_configure'?).

The example which link you sent (Stackoverflow) is using tkinter (not customtkinter). I know that tkinter does the trick, but as I said, I like the nice and neat style of ctk much better,

**Reply by Lautarob1** [reply]
I updated ctk about 2 week ago, is there a new recent version?

**Reply by zSynctic** [reply]
Yeah I know instead of using tag_configure used tag_config this works the stackoverflow solution I sent uses tkinter I just told you to do the same thing since its pretty much the same but using ctk other than that some syntax changes so use ctk

**Reply by zSynctic** [reply]
If your happy with the answer, can you mark it as an answer if you don't mind. thanks

**Comment by Lautarob1** [comment]
Sorry, still doesn’t work:

 

This is the code:

 

        self.txt_Usageinfo = ctk.CTkTextbox(self.frame_content0, wrap=tk.WORD, width=550, height=50)

        self.txt_Usageinfo.grid(row = 6, column = 0, columnspan=3, padx = (50,10), pady=(5,10)) #, sticky = 'w'

 

        lin1 = "Important Usage Information"

        self.txt_Usageinfo.tag_add("lin1","1.0", "1.end")

        self.txt_Usageinfo.tag_config("lin1", font=("Arial", 16, "bold"))

 

        lin2 = "Second line of this test  about usage info"

        self.txt_Usageinfo.tag_add("lin2","3.0", "3.end")

        self.txt_Usageinfo.tag_config("lin2", font=("Arial", 12, "Italic"))

        

        self.txt_Usageinfo.insert("end", lin1+"\n")

        self.txt_Usageinfo.insert("end", lin2+"\n")

 

And this is the error:

    raise AttributeError("'font' option forbidden, because would be incompatible with scaling")

AttributeError: 'font' option forbidden, because would be incompatible with scaling

 

If you have test it, please copy the code snip so I can see where is the error in my code…

 

Thanks a lot!

**Reply by zSynctic** [reply]
Can you provide with me the full code so I can test on it if you want to send it privately u can send it to me in my discord, here is my account: Synctic or synctic.

**Comment by BackspaceVP** [comment]
Still no solution found?
font in ctk. CTkTextbox still doesn't work.

**Reply by Omer-Amin** [reply]
See my answer for a solution.

**Comment by Omer-Amin** [comment]
There is a solution but it's not ideal. You need to overwrite the `tag_config` method to not throw an error when `font` is used. Include this at the start of your code:

```python
def new_tag_config(self, tagName, **kwargs):
        return self._textbox.tag_config(tagName, **kwargs)
customtkinter.CTkTextbox.tag_config = new_tag_config
```

You can then use `tag_config` as usual along with font. Do note that it is not compatible with scaling though, which will produce some unexpected visuals.
